aten/src/ATen/core/dispatch/OperatorEntry.cpp

```diff
@@ -181,23 +184,42 @@ std::pair<const AnnotatedKernel&, const char*> OperatorEntry::computeDispatchTab
     TORCH_INTERNAL_ASSERT(kern_it->second.front().kernel.isValid());
     return {kern_it->second.front(), "kernel"};
 
-  } else if (isIncludedInAlias(dispatch_key, DispatchKey::Autograd)) {
-    // 2.1. For autograd backend keys, use kernel from DispatchKey::Autograd if available
+  }
+
+  // 2.1. Use Math kernel if available. For autograd keys, we only use kernel from Math
+  //      when there's no direct registeration to its corresponding backend key.
+  if (isIncludedInAlias(dispatch_key, DispatchKey::Math)) {
+    auto kern_math = kernels_.find(DispatchKey::Math);
+    if (kern_math != kernels_.end()) {
+      TORCH_INTERNAL_ASSERT(!kern_math->second.empty());
+      TORCH_INTERNAL_ASSERT(kern_math->second.front().kernel.isValid());
+      if (!isIncludedInAlias(dispatch_key, DispatchKey::Autograd)
+          || kernels_.find(getBackendKeyFromAutograd(dispatch_key)) == kernels_.end()) {
+        return {kern_math->second.front(), "math kernel"};
+      }
+    }
+  }
+
+  if (isIncludedInAlias(dispatch_key, DispatchKey::Autograd)) {
+    // 2.2. For autograd backend keys, use kernel from DispatchKey::Autograd if available
     auto kern_autograd = kernels_.find(DispatchKey::Autograd);
     if (kern_autograd != kernels_.end()) {
       TORCH_INTERNAL_ASSERT(!kern_autograd->second.empty());
       TORCH_INTERNAL_ASSERT(kern_autograd->second.front().kernel.isValid());
```

Comment:
Wondering if we should factor this "lookup and then retrieve second.front()" into a helper function of some sort, it's showing up a lot here.